File: 2023/kotlin/day01a/day01a.kt
import java.io.File

fun usage() {
    println("usage: java -jar <jarfile> <infile>")
    System.exit(1)
}

fun process(contents: String): Int {
    var sum = 0
    val digits = listOf("0", "1", "2", "3", "4", "5", "6", "7", "8", "9")
    for ( line in contents.lines() ) {
        var minIndex: Int? = null
        var maxIndex: Int? = null
        var leftDigit: Char? = null
        var rightDigit: Char? = null
        if ( line.isEmpty() ) {
            break
        }
        digits.forEach { digit ->
            if ( line.contains(digit) ) {
                line.indexOf(digit).let { leftIndex ->
                    if (minIndex == null || leftIndex < minIndex!!) {
                        minIndex = leftIndex
                        leftDigit = digit.first()
                    }
                }
                line.lastIndexOf(digit).let { rightIndex ->
                    if (maxIndex == null || rightIndex > maxIndex!!) {
                        maxIndex = rightIndex
                        rightDigit = digit.first()
                    }
                }
            }
        }
        sum += (((leftDigit!!.code) - ('0'.code)) * 10) + ((rightDigit!!.code) - ('0'.code))
    }
    return sum
}

fun main(args: Array<String>) {
    if ( args.size < 1 ) {
        usage()
    }
    val filename = args[0]
    val contents = File(filename).readText()
    val result = process(contents)
    println("result = $result")
}

File: 2023/javascript/day01a/day01a.js
const fs = require('fs');

function usage() {
    console.log(`usage: ${process.argv[0]} ${process.argv[1]} <file>`);
    process.exit(1);
}

function processInput(contents) {
    let lines = contents.trim().split('\n');
    let digits = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];
    let result = 0;
    for ( let line of lines ) {
        let minIndex = null;
        let maxIndex = null;
        let leftDigit = null;
        let rightDigit = null;
        for ( let digit of digits ) {
            if ( line.includes(digit) ) {
                let leftIndex = line.indexOf(digit);
                if ( minIndex == null || leftIndex < minIndex ) {
                    minIndex = leftIndex;
                    leftDigit = digit;
                }
                let rightIndex = line.lastIndexOf(digit);
                if ( maxIndex == null || rightIndex > maxIndex ) {
                    maxIndex = rightIndex;
                    rightDigit = digit;
                }
            }
        }
        result += parseInt(leftDigit + rightDigit);
    }
    return result;
}

function main() {
    if ( process.argv.length < 3 ) {
        usage();
    }
    let filename = process.argv[2];
    let contents = fs.readFileSync(filename, 'utf8');
    let result = processInput(contents);
    console.log(`result = ${result}`);
}

main()
